mypy/newsemanal/semanal_main.py

```diff
@@ -0,0 +1,79 @@
+"""Top-level logic for the new semantic analyzer."""
+
+from typing import List
+
+from mypy.nodes import Node, SymbolTable
+
+MYPY = False
+if MYPY:
+    from mypy.build import Graph, State
+
+
+def semantic_analysis_for_scc(graph: 'Graph', scc: List[str]) -> None:
+    # Assume reachability analysis has already been performed.
+    process_top_levels(graph, scc)
+    process_functions(graph, scc)
+
+
+def process_top_levels(graph: 'Graph', scc: List[str]) -> None:
+    # Process top levels until everything has been bound.
+    # TODO: Limit the number of iterations
+
+    # Initialize ASTs and symbol tables.
+    for id in scc:
+        state = graph[id]
+        assert state.tree is not None
+        state.manager.new_semantic_analyzer.prepare_file(state.tree)
+
+    # Initially all namespaces in the SCC are incomplete (well they are empty).
+    state.manager.incomplete_namespaces.update(scc)
+
+    worklist = scc[:]
+    while worklist:
+        deferred = []  # type: List[str]
+        while worklist:
+            next_id = worklist.pop()
+            deferred += semantic_analyze_target(next_id, graph[next_id])
+            # Assume this namespace is ready.
+            # TODO: It could still be incomplete if some definitions couldn't be bound.
+            state.manager.incomplete_namespaces.discard(next_id)
+        worklist = deferred
+
+
+def process_functions(graph: 'Graph', scc: List[str]) -> None:
+    # TODO: This doesn't quite work yet
+    # Process functions.
+    deferred = []  # type: List[str]
+    for id in scc:
+        tree = graph[id].tree
+        assert tree is not None
+        symtable = tree.names
+        targets = get_all_leaf_targets(symtable)
+        for target in targets:
+            deferred += semantic_analyze_target(target, graph[id])
+    assert not deferred  # There can't be cross-function forward refs
```

Comment:
It's not a limitation as such, since deferring functions should never be necessary. All namespaces need to complete before start processing functions.